### PR TITLE
Migration to Laminas

### PR DESCRIPTION
**Pull request by snapshotpl**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -22,9 +22,10 @@
         "php": "^7.1",
         "container-interop/container-interop": "^1.0",
         "monolog/monolog": "^1.3 || ^2.0",
-        "zendframework/zend-modulemanager": "^2",
-        "zendframework/zend-stdlib": "^2 || ^3",
-        "zendframework/zend-servicemanager": "^2 || ^3"
+        "laminas/laminas-modulemanager": "^2",
+        "laminas/laminas-stdlib": "^2 || ^3",
+        "laminas/laminas-servicemanager": "^2 || ^3",
+        "laminas/laminas-dependency-plugin": "^1.0"
```
Comment (michalbundyra) [inline composer.json]: It should not be used in external libraries, when we migrate to Laminas.

**Comment by michalbundyra** [comment]
@snapshotpl I've reported a bug ticket for composer, here: https://github.com/composer/composer/issues/8677.

**Comment by abacaphiliac** [comment]
@snapshotpl this is great.

i was assuming this migration would require a v3 release, as this project is no longer a Zend integration, it's a Laminas integration. if we go to v3 then we can finally drop Zend v2 and make some other breaking changes we've been waiting for.

if you feel strongly that we should keep Laminas v2, please tell me why.

**Comment by snapshotpl** [comment]
@michalbundyra Is it fixed now https://github.com/composer/composer/issues/8677#issuecomment-597018632 ?

@abacaphiliac I personally don't need v2, however it will be nice to keep this compatibility if possible since v2 is still supported. Have I right @michalbundyra ?

**Comment by abacaphiliac** [comment]
@snapshotpl i've been thinking about this a lot. the way i see it, we have 3 options. let's explore.

1. migrate in version 2
2. migrate in version 3
3. create a new project for this new integration between monolog and laminas

let's address number 3 first because i like it the least, haha:

ever since i heard that zendframework would be archived and superceded by laminas i had planned to do the migration in this project. i don't feel this project should die just because one of it's dependencies has been copied into a new namespace and the old one archived.

migrate in version 2:

i think this will break consumers. suppose i have an application which uses `"enlitepro/enlite-monolog": "^2"`. it also explicitly depends upon zendframework 2 components shared by this project, for example `"zendframework/zendframework": "^2"` (i know this is "old hat", but it is simpler for this exercise and i think it will suffice). if/when we migrate to laminas in the way you've proposed here and release in version 2 then composer will happily upgrade my application to use the migrated version. this means my application will still have zendframework components due to explicit dependency and will now have a few laminas dependencies. having extra packages isn't a big deal, but the application is configuring and running zendframework, not laminas. so i think we will have a runtime conflict, not an install/build-time conflict, where the service-manager and/or module-manager don't work the way we'd expect with respect to logging and this integration. we can explore this further, if i haven't convinced you. maybe we could come up with a test case. i'd be happy to spend time on that before potentially doing something that i think will break our users.

migrate in version 3:

this is the safest route. there will be clear line for users to cross. we can provide migration documentation, etc. the only remaining issue though is that users could still install version 3 of this project (with laminas integration) alongside archived zendframework components. for that reason i feel that we should also add a [conflict](https://getcomposer.org/doc/04-schema.md#conflict) config  to prevent that from happening. then our users will get an early warning that this project only works with laminas moving forward.

notes:

we haven't discussed a true "compatibility" release, where this project works with both zendframework and laminas, instead of one or the other. this will double the source code and double the test cases (if we truly want to support laminas v2, for example). some may advocate for this, but i'm not one of them. i'm not sure the juice is worth the squeeze. let's debate, though, if you feel strongly.

i think it's also worth mentioning that we could add a [conflict](https://getcomposer.org/doc/04-schema.md#conflict) config in a version 2 migration, but i feel that will just add confusion even though the break will be caught before runtime (at install).

**Comment by boesing** [comment]
@abacaphiliac thanks for pointing out the possibilities. I'd like to vote for option 3 aswell.
When monolog would change its name (as zendframework did for laminas), this package should migrate to a new one due to visibility.

Thats what I am doing with https://github.com/boesing/zend-cache-redis-cluster aswell.

---

I am actually migrating to laminas and see that this repository is not yet migrated.
Thats not a problem for now, as there are no new major versions required so far.

Is there anything we can help to get this merged into v3?

Thanks in advance!

---

Regarding the notes: this package is compatible with both laminas and zendframework as laminas packages provide a bridge and thus, `Zend\` namespace objects are still "available" (but aliased to the laminas equivalents). There is no need for any 3rd-party library to have that kind of functionality. 👍🏼

**Comment by snapshotpl** [comment]
@abacaphiliac ok, let's do it! Do you need help with this?

**Comment by abacaphiliac** [comment]
@boesing thank you for the feedback. i agree 100% that bridge functionality is the responsibility of the framework and that they have provided a migration path via additional optional packages that can be installed to take advantage of aliases. i wasn't familiar with the `replace` configuration in Composer, so thank you for that as well. it does seem there is some confusion with how it should behave and i was proposing `conflict` so that application developers wouldn't accidentally install ZendFramework and Laminas side-by-side in the same way that [SecurityAdvisories](https://github.com/Roave/SecurityAdvisories/blob/latest/composer.json#L18) prevents installation of "incompatible" packages for the purposes of security. as you pointed out, that's not the responsibility of a third-party library so i think we're in alignment here and shouldn't worry too much about being everything to everyone all of the time.

@snapshotpl yes, please. i think we need only two things to get this merged in: we need to drop the ZendFramework code, and we need to add v3 as a default branch alias in the Composer config.

**Comment by boesing** [comment]
@abacaphiliac I think `laminas-ci` could be of help here, even tho, it is a laminas related project 👍🏼 
If you are interested in how that works, have a look in the matrix action:

https://github.com/laminas/laminas-ci-matrix-action

Here is a result of a running CI pipeline in mezzio: https://github.com/mezzio/mezzio/actions/runs/861276545

Migration to that is quite straight forward (even tho some repo tokens might be necessary), there is a step-by-step guide from matthew: 
https://gist.github.com/weierophinney/b003e50c3c2667d08076caf31ebd36a4

More details on the CI action execution can be found here:
https://github.com/laminas/laminas-continuous-integration-action

(seems like the `GITHUB_TOKEN` is not necessarily required: https://github.com/laminas/laminas-continuous-integration-action#github_token-not-required)

**Comment by snapshotpl** [comment]
Again problem with Scrutinizer... Also travis doesn't work, probably they moved from org to com

**Comment by snapshotpl** [comment]
@abacaphiliac probably you need to change something in settings to make it work again

**Comment by abacaphiliac** [comment]
@snapshotpl i'm on it. i'm replacing Travis CI with GitHub Actions in https://github.com/enlitepro/enlite-monolog/pull/69. not sure what's up with the Scrutinizer integration but i'm looking into it.

**Comment by abacaphiliac** [comment]
@snapshotpl ok, i think i have something that works in #69. would you mind please taking a look and making sure there's no funny business? ; )

**Comment by abacaphiliac** [comment]
@snapshotpl thanks so much for the review on #69. i'm hoping that this PR will be able to build successfully again after you rebase it. please let me know how that goes or if i can help in any other way.

**Comment by snapshotpl** [comment]
@abacaphiliac done!

**Comment by boesing** [comment]
As this might become a new major release, I have a few things to consider:

- removal of the `modulemanager` dependency - we are using this component in a mezzio project which does not use the `laminas/laminas-modulemanager` component but the `laminas/laminas-config-aggregator` component - **any `Module` class containing a public `getConfig` method is totally fine for `laminas-modulemanager`!** (see [laminas/laminas-cache](https://github.com/laminas/laminas-cache/blob/2.12.x/composer.json) for example)
- support for [laminas/laminas-config-aggregator](https://github.com/laminas/laminas-config-aggregator) by moving all config related stuff to a new `ConfigProvider` class (while using its methods in the `Module` class to keep support for modulemanager)
- support for [laminas/laminas-component-installer](https://github.com/laminas/laminas-component-installer) (can be done in any minor aswell) - can provide a `component` as well as a `config-provider` section under `extra.laminas` and thus enables auto-detection during `composer install/require/uninstall/update`

**Comment by boesing** [comment]
Awesome, thanks @snapshotpl!

**Review by boesing — APPROVED** [review]
LGTM

**Comment by snapshotpl** [comment]
@abacaphiliac waiting for your approvement :-)

File: src/EnliteMonolog/Service/MonologServiceAwareInterface.php [hunk 2]
```diff
@@ -9,15 +9,7 @@
...
-     * @param Logger $monologService
-     * @return void
-     */
-    public function setMonologService(Logger $monologService);
-
-    /**
-     * @return Logger
-     */
-    public function getMonologService();
+    public function getMonologService(): ?Logger;
```
Comment (abacaphiliac) [inline src/EnliteMonolog/Service/MonologServiceAwareInterface.php:14]: why nullable? when would an implementation of this interface not return a Logger from this method?

**Review by abacaphiliac — COMMENTED** [review]
@snapshotpl really nice job here. thanks for sticking with it. just one quick question on this proposal (possibly a change request), and a short suggestion to squeeze in one more PR based on a commit from #71.. if you're willing.

@boesing thank you so much for jumping in with additional suggestions and expertise.

**Comment by boesing** [comment]
@abacaphiliac you're welcome!

In laminas, there can be plenty of plugin managers. there is no way to specify an initializer for **all** of them without knowing the config key.
So every project can use the aware Interface but could not add the initializer to a specific plugin manager. So just because you implement an Interface does not mean that everything works. That is why initializers should be avoided and instead, constructor injection is suggested over setter injection.

If you want a safe way to always have a logger in your instance, use construct injection.
If you are fine with not having a logger, use setter injection. 

For example, there are about 6 plugin managers in laminas.

- Filters
- validators
- Controller (MVC)
- cache (storage Adapter)
- views (laminas-view)
- ... (there are more - you might want to check laminas-config-aggregator-modulemanager to verify, I woke up recently and dont have it in my head right now 😅)

but this Module only assigns the initializer to the servicemanager only. So all validators implementing this Interface will never receive the service.

File: src/EnliteMonolog/Service/MonologServiceAwareInterface.php (hunk 2 above)
Comment (boesing) [reply src/EnliteMonolog/Service/MonologServiceAwareInterface.php:14]: Oh, my [comment](https://github.com/enlitepro/enlite-monolog/pull/68#issuecomment-871906081) references this question.

**Comment by abacaphiliac** [comment]
@boesing thank you for the thoughtful explanation. i am on the same page regarding initializers (and traits) versus constructor injection, for some time now. i suppose my nullable callout was more about the probably incomplete doc-block currently in the main branch. i can certainly buy that the docs were wrong before and the return type is now accurate and forces consumers to deal with the reality of something not being present.

what are your thoughts on keeping the initializer and trait around? is it old-hat enough that they should be dropped? in a major version? has the ecosystem moved on? or should we leave them in to reduce the amount of breaking changes?

**Comment by boesing** [comment]
I'd say keep it for now and mark it deprecated. 
Kick it in the next major even tho this might take quite a long time for new bc breaks. 

But maintaining it a little longer should be doable 👍🏻

**Comment by snapshotpl** [comment]
Can we add more changes in another PRs? :-)

**Comment by abacaphiliac** [comment]
awesome work @snapshotpl, this has been coming for a long time and you made it happen. thank you.

**Comment by snapshotpl** [comment]
@abacaphiliac awesome! What next? I want to make it released with PHP 8 support because it's last peace to update my stuff.

**Comment by abacaphiliac** [comment]
@snapshotpl thank you for your patience. [3.0.0](https://packagist.org/packages/enlitepro/enlite-monolog#3.0.0) has been published. thank you for all of your hard work and dedication to get this milestone over the finish line.

**Pull request merged**